src/session/session_prefetch.c

```diff
@@ -10,55 +10,53 @@
+ *     Check if pre-fetching work should be performed for a given ref.
  */
 bool
 __wt_session_prefetch_check(WT_SESSION_IMPL *session, WT_REF *ref)
 {
-    /* Internal threads should not be configured to do pre-fetching. */
-    if (!S2C(session)->prefetch_auto_on || F_ISSET(session, WT_SESSION_INTERNAL))
-        return (false);
-
     if (S2C(session)->prefetch_queue_count > WT_MAX_PREFETCH_QUEUE)
         return (false);
-
     /*
-     * Don't deal with internal pages at the moment - finding the right content to preload based on
-     * internal pages is hard.
+     * Check if pre-fetching is enabled on the session level. We don't perform pre-fetching on
+     * internal threads or internal pages (finding the right content to preload based on internal
+     * pages is hard), so check for that too. We also want to pre-fetch pages that have had at least
+     * one read. The result of this check will be used by cursor logic to determine if pre-fetching
+     * will be performed.
      */
-    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
-        return (false);
-
-    if (session->pf.prefetch_disk_read_count == 1)
-        WT_STAT_CONN_INCR(session, block_prefetch_disk_one);
-
-    /* A single read from disk is common - don't use it to guide pre-fetch behavior. */
-    if (session->pf.prefetch_disk_read_count < 2) {
+    if (!F_ISSET(session, WT_SESSION_PREFETCH) || F_ISSET(session, WT_SESSION_INTERNAL) ||
+      F_ISSET(ref, WT_REF_FLAG_INTERNAL) || session->pf.prefetch_disk_read_count < 2) {
         WT_STAT_CONN_INCR(session, block_prefetch_skipped);
         return (false);
     }
 
-    if (session->pf.prefetch_prev_ref == NULL) {
-        WT_STAT_CONN_INCR(session, block_prefetch_attempts);
-        return (true);
-    }
-
     /*
-     * If the previous pre-fetch was using the same home ref, pre-fetch for approximately the number
-     * of pages that were added to the queue.
+     * We want to avoid the scenario of requesting pre-fetch on one particular ref many times (e.g
+     * when reading along a single page). We can identify this by checking if the previous pre-fetch
+     * was performed using the same home ref.
+     *
+     * In the event that we find this to be true, we perform pre-fetch for approximately the number
+     * of pages that were added to the queue (WT_PREFETCH_QUEUE_PER_TRIGGER). We then want to ensure
+     * that we will not pre-fetch from this ref for a while, and this is done by checking a counter.
+     *
+     * The counter variable prefetch_skipped_with_parent tracks the number of skips we have
+     * performed on a particular ref. If the number of skips surpasses the number of pages that have
+     * been queued for pre-fetch, we are okay to pre-fetch from this ref again. This condition will
+     * evaluate to false and the counter will be reset, effectively marking the ref as available to
+     * pre-fetch from.
      */
     if (session->pf.prefetch_prev_ref->page == ref->home &&
```

Comment:
We're dereferencing this without a null check - has something else changed (e.g. in the last PR) to make this OK now?

Reply:
I didn't do a null check for this as I believed that `prefetch_prev_ref` could not be NULL. I see it being set in only one place `__wt_btree_prefetch` as `session->pf.prefetch_prev_ref = ref;` where ref is meant to always have a valid page as quoted by the function comment:

"This session holds a hazard pointer on the ref passed in, so there must be a valid page and a valid parent page"

Happy to still include a null check as a safeguard for the future though. Let me know your thoughts!

Reply:
Thanks for taking the time to fact-check me on this - I believe you're right. If it's a precondition for the function, let's add it as an assertion?